### PR TITLE
1398 androidx update

### LINKED ISSUE #1398: Move support libraries to androidx
### Why/User Benefit/User Problem
Like #991, we have to stay up to date with the latest android conventions. Staying up to date ensures we can add new android features for users easily. This migration is part of that. I don't know if there's a date that we have to do this by but I'd assume it's by next fall, for SDK 29. That being said, components and other libraries that migrate may force us to adopt this so we should do it sooner.

### What / Requirements
- Use migration tool to move to new support libraries
- Use jetifier to update dependencies to use androidx too
- Add new toys that we're unable to access like androidx KTX extensions (see [https://developer.android.com/topic/libraries/architecture/adding-components](https://developer.android.com/topic/libraries/architecture/adding-components)

### Acceptance Criteria (how do I know when I’m done?)
- QA acceptance that there are no regressions
  - Focus on dependency integrations (b/c jetifier rewrites)

### PR DESCRIPTION
**Pull request by Baron-Severin**
Fixes #1398

### Pull Request checklist

- [X] This PR includes thorough **tests** or an explanation of why it does not
Dependency change only, should be nonfunctional
- [x] This PR includes a **CHANGELOG entry** or does not need one
- [ ] The **UI tests** are passing after this PR (`./gradlew connectedSystemDebugAndroidTest`)
`DOMElementFocusTest` is failing, but it seems to be failing on master as well
- [X] I have considered adding **QA labels** on the associated issue (not this PR; `qa-ready` or `qa-denied`)

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -152,7 +145,7 @@ dependencies {
 
     systemImplementation "org.mozilla.components:browser-engine-system:$moz_components_version"
     // Required to override android.support lib in GeckoView dependency (uses palette-v7:26.1.0)
-    geckoImplementation "com.android.support:palette-v7:$support_libraries_version"
+    geckoImplementation "androidx.palette:palette:$support_libraries_version"
```
Comment (psymoon) [inline app/build.gradle]: Did you check gecko variant's correctly rendering web pages? :)

File: app/build.gradle [hunk 2]
```diff
@@ -114,22 +114,15 @@ configurations {
...
-    }
+    implementation "androidx.appcompat:appcompat:$support_libraries_version"
+    implementation "androidx.legacy:legacy-support-v4:$support_libraries_version"
+    implementation "androidx.cardview:cardview:$support_libraries_version"
+    implementation "androidx.browser:browser:$support_libraries_version"
+    implementation "com.google.android.material:material:$support_libraries_version"
+    implementation "androidx.leanback:leanback:$support_libraries_version"
+    implementation "androidx.recyclerview:recyclerview:$support_libraries_version"
+    implementation "androidx.constraintlayout:constraintlayout:1.1.3"
+    implementation 'com.squareup.picasso:picasso:2.71828'
```
Comment (psymoon) [inline app/build.gradle]: So we don't need to exclude its parent dependencies anymore?

File: app/build.gradle [hunk 3]
```diff
@@ -114,22 +114,15 @@ configurations {
...
-        exclude group: 'com.android.support', module: 'exifinterface'
-        exclude group: 'com.android.support', module: 'support-annotations'
-        exclude group: 'com.android.support', module: 'support-compat'
-        exclude group: 'com.android.support', module: 'support-fragment'
-    }
+    implementation "androidx.appcompat:appcompat:$support_libraries_version"
+    implementation "androidx.legacy:legacy-support-v4:$support_libraries_version"
+    implementation "androidx.cardview:cardview:$support_libraries_version"
+    implementation "androidx.browser:browser:$support_libraries_version"
+    implementation "com.google.android.material:material:$support_libraries_version"
```
Comment (psymoon) [inline app/build.gradle]: This looks like it's new, do we need this?

File: app/src/main/java/org/mozilla/tv/firefox/architecture/FirefoxViewModelProviders.kt [hunk 4]
```diff
@@ -17,11 +17,11 @@ import org.mozilla.tv.firefox.ext.serviceLocator
  */
 object FirefoxViewModelProviders {
 
-    fun of(activity: FragmentActivity): ViewModelProvider {
+    fun of(activity: androidx.fragment.app.FragmentActivity): ViewModelProvider {
```
Comment (psymoon) [inline app/src/main/java/org/mozilla/tv/firefox/architecture/FirefoxViewModelProviders.kt]: Looks like a Android Studio build residue. I think this happens if there are more declarations of same class names from different packages - which shouldn't be the case anymore? Try clean, rebuild and re-import?

File: app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/ToolbarUiController.kt [hunk 5]
```diff
@@ -36,7 +36,7 @@ class ToolbarUiController(
 
     private var hasUserChangedURLSinceEditTextFocused = false
 
-    fun onCreateView(layout: View, viewLifecycleOwner: LifecycleOwner, fragmentManager: FragmentManager) {
+    fun onCreateView(layout: View, viewLifecycleOwner: LifecycleOwner, fragmentManager: androidx.fragment.app.FragmentManager) {
```
Comment (psymoon) [inline app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/ToolbarUiController.kt]: See above for AS build residue

File: app/src/main/java/org/mozilla/tv/firefox/pinnedtile/PinnedTileAdapter.kt [hunk 6]
```diff
@@ -44,7 +44,7 @@ class PinnedTileAdapter(
     private val loadUrl: (String) -> Unit,
     val onTileLongClick: (() -> Unit)?,
     val onTileFocused: (() -> Unit)?
-) : RecyclerView.Adapter<TileViewHolder>() {
+) : androidx.recyclerview.widget.RecyclerView.Adapter<TileViewHolder>() {
```
Comment (psymoon) [inline app/src/main/java/org/mozilla/tv/firefox/pinnedtile/PinnedTileAdapter.kt]: See above for AS build residue

File: app/src/main/java/org/mozilla/tv/firefox/webrender/VideoVoiceCommandMediaSession.kt [hunk 7]
```diff
@@ -317,7 +321,7 @@ class VideoVoiceCommandMediaSession @UiThread constructor(
                 return true
             }
 
-            return super.onMediaButtonEvent(mediaButtonEvent)
+            return super.onMediaButtonEvent(mediaButtonEvent!!)
```
Comment (psymoon) [inline app/src/main/java/org/mozilla/tv/firefox/webrender/VideoVoiceCommandMediaSession.kt]: Is mediaButtonEvent instance garuntee'd here?

File: app/src/main/java/org/mozilla/tv/firefox/webrender/LocalizedContent.kt [hunk 8]
```diff
@@ -21,7 +21,7 @@ object LocalizedContent {
     fun generateAboutPage(context: Context): String {
         val resources = Locales.getLocalizedResources(context)
 
-        val substitutionMap = ArrayMap<String, String>()
+        val substitutionMap = androidx.collection.ArrayMap<String, String>()
```
Comment (psymoon) [inline app/src/main/java/org/mozilla/tv/firefox/webrender/LocalizedContent.kt]: See above for AS build residue

File: app/src/main/res/layout/activity_onboarding.xml [hunk 9]
```diff
@@ -2,15 +2,15 @@
 <!-- This Source Code Form is subject to the terms of the Mozilla Public
    - License, v. 2.0. If a copy of the MPL was not distributed with this
    - file, You can obtain one at http://mozilla.org/MPL/2.0/. -->
-<android.support.design.widget.CoordinatorLayout xmlns:android="http://schemas.android.com/apk/res/android"
-    xmlns:tools="http://schemas.android.com/tools"
-    android:layout_width="match_parent"
-    android:layout_height="match_parent"
-    tools:context="org.mozilla.tv.firefox.onboarding.OnboardingActivity">
+<androidx.coordinatorlayout.widget.CoordinatorLayout xmlns:android="http://schemas.android.com/apk/res/android"
+                                                     xmlns:tools="http://schemas.android.com/tools"
```
Comment (psymoon) [inline app/src/main/res/layout/activity_onboarding.xml]: Is the indentation spacing intentional here?

File: app/src/main/res/layout/content_onboarding.xml [hunk 10]
```diff
@@ -1,21 +1,21 @@
...
-<android.support.constraint.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
-    xmlns:app="http://schemas.android.com/apk/res-auto"
-    xmlns:tools="http://schemas.android.com/tools"
-    android:layout_width="fill_parent"
-    android:layout_height="fill_parent"
-    app:layout_behavior="@string/appbar_scrolling_view_behavior"
-    tools:context="org.mozilla.tv.firefox.onboarding.OnboardingActivity"
-    tools:showIn="@layout/activity_onboarding">
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+                                                   xmlns:app="http://schemas.android.com/apk/res-auto"
```
Comment (psymoon) [inline app/src/main/res/layout/content_onboarding.xml]: nit: indentation spacing - see above

File: app/src/main/res/layout/fragment_navigation_overlay.xml [hunk 11]
```diff
@@ -1,15 +1,15 @@
...
-    xmlns:tools="http://schemas.android.com/tools"
-    xmlns:app="http://schemas.android.com/apk/res-auto"
-    android:id="@+id/overlayScrollView"
-    android:layout_width="match_parent"
-    android:layout_height="match_parent"
-    android:fillViewport="true"
-    android:focusable="false"
-    android:scrollbars="none">
+<androidx.coordinatorlayout.widget.CoordinatorLayout xmlns:android="http://schemas.android.com/apk/res/android"
+                                                     xmlns:tools="http://schemas.android.com/tools"
```
Comment (psymoon) [inline app/src/main/res/layout/fragment_navigation_overlay.xml]: nit: indentation spacing - see above

File: app/src/main/res/layout/fragment_pocket_video.xml [hunk 12]
```diff
@@ -4,21 +4,21 @@
...
 <!-- TODO: figure out how to handle Overdraw in #849 -->
-<android.support.constraint.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
-    xmlns:app="http://schemas.android.com/apk/res-auto"
-    xmlns:tools="http://schemas.android.com/tools"
-    android:layout_width="match_parent"
-    android:layout_height="match_parent"
-    android:background="@color/photonGrey70"
-    tools:ignore="Overdraw"
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+                                                   xmlns:app="http://schemas.android.com/apk/res-auto"
```
Comment (psymoon) [inline app/src/main/res/layout/fragment_pocket_video.xml]: nit: indentation spacing - see above

File: build.gradle [hunk 13]
```diff
@@ -3,23 +3,23 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 buildscript {
-    ext.support_libraries_version = '28.0.0'
-    ext.architecture_components_version = '1.1.1'
+    ext.support_libraries_version = '1.0.0'
```
Comment (psymoon) [inline build.gradle]: Should this be rename this to `androidx_version`? RIP `support_libraries`

File: build.gradle [hunk 14]
```diff
@@ -3,23 +3,23 @@
...
+    ext.architecture_components_version = '2.0.0'
     ext.kotlin_version = '1.3.0'
     ext.moz_components_version = '0.43.0'
 
-    ext.robolectric_version = '4.0.1' // required for SDK 28.
+    ext.robolectric_version = '4.0.2' // required for SDK 28.
```
Comment (psymoon) [inline build.gradle]: nit: Add a comment for version bump reason - androidx?

File: app/build.gradle (hunk 2 above)
Comment (Baron-Severin) [reply app/build.gradle]: The comment said that this was blocked on #883, which was closed a while ago.  I removed them and haven't seen any problems, so I think we're good here.

File: app/build.gradle (hunk 3 above)
Comment (Baron-Severin) [reply app/build.gradle]: This is the new name for `com.android.support:design`.  Crazy, right?

File: app/src/main/res/layout/activity_onboarding.xml (hunk 9 above)
Comment (Baron-Severin) [reply app/src/main/res/layout/activity_onboarding.xml]: Nope, fixed.

File: app/src/main/java/org/mozilla/tv/firefox/webrender/VideoVoiceCommandMediaSession.kt (hunk 7 above)
Comment (Baron-Severin) [reply app/src/main/java/org/mozilla/tv/firefox/webrender/VideoVoiceCommandMediaSession.kt]: **Thank you for catching this!**  I meant to put a TODO on that and look into it later, and completely forgot.  Really bad mistake on my part.

File: build.gradle (hunk 14 above)
Comment (Baron-Severin) [reply build.gradle]: I should have when I did it.  I know there was a good reason for bumping this, but I honestly can't remember what it was.

File: app/build.gradle (hunk 1 above)
Comment (Baron-Severin) [reply app/build.gradle]: I hadn't!  Seems to work ok, aside from the known issues.  Is there anything in particular you would like me to look at?

**Pull request merged**